### PR TITLE
ADBDEV-4938: Add assertions to return from dynamic_cast.

### PR DESCRIPTION
**Pull request by RekGRpth**
Add assertions to return from dynamic_cast.

dynamic_cast may return NULL, so I added assertions.

In the CParseHandlerLogicalCTAS::EndElement and
CParseHandlerMDRelationCtas::EndElement methods, dynamic_cast has been removed
for the variables opfamilies_parse_handler and opclasses_parse_handler, which
already have the corresponding type CParseHandlerMetadataIdList*.

### CODE REVIEW

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerLogicalCTAS.cpp [hunk 1]
```diff
@@ -232,11 +238,13 @@ CParseHandlerLogicalCTAS::EndElement(const XMLCh *const,  // element_uri,
 	IMdIdArray *distr_opfamilies =
 		dynamic_cast<CParseHandlerMetadataIdList *>(opfamilies_parse_handler)
 			->GetMdIdArray();
+	GPOS_ASSERT(NULL != distr_opfamilies);
 	distr_opfamilies->AddRef();
 
 	IMdIdArray *distr_opclasses =
 		dynamic_cast<CParseHandlerMetadataIdList *>(opclasses_parse_handler)
 			->GetMdIdArray();
```
Comment (KnightMurloc) [inline src/backend/gporca/libnaucrates/src/parser/CParseHandlerLogicalCTAS.cpp]: The dynamic_cast result is accessed on the same line. To check, you need to make the casting in a separate variable. The same thing is above.
```suggestion
   CParseHandlerMetadataIdList* metadata = dynamic_cast<CParseHandlerMetadataIdList *>(opclasses_parse_handler);
   GPOS_ASSERT(NULL != metadata);
	IMdIdArray *distr_opclasses = metadata->GetMdIdArray();
```
By the way are casting is necessary here?

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelationCtas.cpp [hunk 2]
```diff
@@ -192,11 +197,13 @@ CParseHandlerMDRelationCtas::EndElement(const XMLCh *const,	 // element_uri,
...
 		dynamic_cast<CParseHandlerMetadataIdList *>(opfamilies_parse_handler)
 			->GetMdIdArray();
+	GPOS_ASSERT(NULL != distr_opfamilies);
 	distr_opfamilies->AddRef();
 
 	IMdIdArray *distr_opclasses =
 		dynamic_cast<CParseHandlerMetadataIdList *>(opclasses_parse_handler)
 			->GetMdIdArray();
+	GPOS_ASSERT(NULL != distr_opclasses);
 	distr_opclasses->AddRef();
```
Comment (KnightMurloc) [inline src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelationCtas.cpp]: same

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelation.cpp [hunk 3]
```diff
@@ -299,6 +303,7 @@ CParseHandlerMDRelation::EndElement(const XMLCh *const,	 // element_uri,
 	CMDColumnArray *md_col_array = md_cols_parse_handler->GetMdColArray();
 	CMDIndexInfoArray *md_index_info_array =
 		pphMdlIndexInfo->GetMdIndexInfoArray();
+	GPOS_ASSERT(NULL != pphMdidlTriggers);
```
Comment (KnightMurloc) [inline src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelation.cpp]: move this assert to the rest of the group.

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelationExternal.cpp [hunk 4]
```diff
@@ -252,6 +258,7 @@ CParseHandlerMDRelationExternal::EndElement(
 		distr_opfamilies = dynamic_cast<CParseHandlerMetadataIdList *>(
 							   m_opfamilies_parse_handler)
 							   ->GetMdIdArray();
+		GPOS_ASSERT(NULL != distr_opfamilies);
```
Comment (KnightMurloc) [inline src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelationExternal.cpp]: same

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerLogicalCTAS.cpp (hunk 1 above)
Comment (RekGRpth) [reply src/backend/gporca/libnaucrates/src/parser/CParseHandlerLogicalCTAS.cpp]: fixed

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelation.cpp (hunk 3 above)
Comment (RekGRpth) [reply src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelation.cpp]: moved

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelationCtas.cpp (hunk 2 above)
Comment (RekGRpth) [reply src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelationCtas.cpp]: fixed

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelationExternal.cpp (hunk 4 above)
Comment (RekGRpth) [reply src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelationExternal.cpp]: fixed

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerLogicalCTAS.cpp (hunk 1 above)
Comment (KnightMurloc) [reply src/backend/gporca/libnaucrates/src/parser/CParseHandlerLogicalCTAS.cpp]: So why do we need type casting here? It looks like opclasses_parse_handler already has the CParseHandlerMetadataIdList type.

Comment (RekGRpth) [reply src/backend/gporca/libnaucrates/src/parser/CParseHandlerLogicalCTAS.cpp]: indeed, fixed

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerLogicalCTAS.cpp [hunk 5]
```diff
@@ -228,18 +234,11 @@ CParseHandlerLogicalCTAS::EndElement(const XMLCh *const,  // element_uri,
...
+	IMdIdArray *distr_opfamilies = opfamilies_parse_handler->GetMdIdArray();
 	distr_opfamilies->AddRef();
-
-	IMdIdArray *distr_opclasses =
-		dynamic_cast<CParseHandlerMetadataIdList *>(opclasses_parse_handler)
-			->GetMdIdArray();
+	IMdIdArray *distr_opclasses = opclasses_parse_handler->GetMdIdArray();
 	distr_opclasses->AddRef();
 
-
```
Comment (KnightMurloc) [inline src/backend/gporca/libnaucrates/src/parser/CParseHandlerLogicalCTAS.cpp:242]: unnecessary change

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelationCtas.cpp [hunk 6]
```diff
@@ -189,17 +194,11 @@ CParseHandlerMDRelationCtas::EndElement(const XMLCh *const,	 // element_uri,
...
+	IMdIdArray *distr_opfamilies = opfamilies_parse_handler->GetMdIdArray();
 	distr_opfamilies->AddRef();
-
-	IMdIdArray *distr_opclasses =
-		dynamic_cast<CParseHandlerMetadataIdList *>(opclasses_parse_handler)
-			->GetMdIdArray();
+	IMdIdArray *distr_opclasses = opclasses_parse_handler->GetMdIdArray();
 	distr_opclasses->AddRef();
 
-
```
Comment (KnightMurloc) [inline src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelationCtas.cpp:202]: unnecessary change

**Comment by KnightMurloc** [comment]
Add the reasons for deleting dynamic_cast to the description.

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelationCtas.cpp (hunk 6 above)
Comment (RekGRpth) [reply src/backend/gporca/libnaucrates/src/parser/CParseHandlerMDRelationCtas.cpp:202]: reverted

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerLogicalCTAS.cpp (hunk 5 above)
Comment (RekGRpth) [reply src/backend/gporca/libnaucrates/src/parser/CParseHandlerLogicalCTAS.cpp:242]: reverted

**Comment by RekGRpth** [comment]
> Add the reasons for deleting dynamic_cast to the description.

added

**Review by dnskvlnk — APPROVED** [review]

**Review by KnightMurloc — APPROVED** [review]

**Comment by RekGRpth** [comment]
https://github.com/arenadata/gpdb/pull/870
